Remove bootstrap CSS rules for radio buttons.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is to add a CSS file for emails page in dev server. This is required to add css for focus outline as we cannot use inline styling for it.
- Next 4 commits are to individually add bootstrap CSS rules to the relevant files and then one more commit to extract some common rules in `zulip.css`.
- Last commit is to remove the styles for radio buttons from `bootstrap.css`.

Have added some information in commit messages about why a couple of rules are skipped due to them already being default or redundant.

We use radio type inputs in `team` and `upgrade` pages, but the `input` elements are essentially hidden and we just style our own components already, so this change does not effect them.
<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

sahil839 commented:
Also, I do not think the current focus outline looks good especially the ones for buttons for `Enter to send` setting. Note that the behavior is not changed in this PR, so can probably be solved as a follow-up. Sharing some screenshots below for the current behavior -

In Google Chrome -
![Screenshot from 2022-11-23 23-17-10](https://user-images.githubusercontent.com/35494118/203629595-f6a9e754-2440-442e-b829-1bdb0f0f7a1f.png)

![Screenshot from 2022-11-23 23-18-12](https://user-images.githubusercontent.com/35494118/203629644-76e5e215-8487-4866-8c19-c9935089bd23.png)

In Firefox -
![Screenshot from 2022-11-23 23-20-12](https://user-images.githubusercontent.com/35494118/203629673-e960c31b-0608-472d-8d6f-57108c047d99.png)
![Screenshot from 2022-11-23 23-21-05](https://user-images.githubusercontent.com/35494118/203629703-d134f07f-6f86-4a7e-be72-6a18dbb290da.png)

timabbott commented on static/styles/zulip.css:
```diff
@@ -73,6 +73,21 @@ a {
+    &:disabled {
+        cursor: not-allowed;
+    }
+}
```
So a big part of the goal here is not to have `zulip.css` CSS for all radio inputs -- we're trying to replace that CSS style with reusable components, like "settings_checkbox", that have a specific style. 

I think the right model is to have a `input[type="radio"].settings-radio-input` class, and use that in all the settings contexts; and then be OK with a bit of duplication with the one non-settings-styled use case (the enter-sends widget). 

We don't have to go all the way there in this PR, but the direction should be that rather than having CSS for a list of specific settings ratio elements like this:
```
.emojiset_choices,
.user_list_style_values
```

We'd just have CSS for a `settings-radio-input` object, and an HTML partial template that can be included when one wants such a component in the app. Or maybe we end up with two types -- a simple radio input and a `display_settings_radio_input` class that's for the ones where we show a visual preview. 

In particular, we want to eliminate all CSS that applies to all radio inputs, since it means that any radio input that happens to want different styling needs to override the others.

Does that make sense?

timabbott replied:
I think probably this model suggests a different commit structure -- a commit could introduce a new common class and use it in one place, and later commits could use it in more places; or maybe even better given there's only 2 settings radio buttons with the preview, it'd probably be fine to just do a commit that makes the class and changes both of them to use it. And then another for variants.

sahil839 replied:
For the CSS in `zulip.css` my thinking is that atleast the `cursor` property would be constant across the app (and width will probably be removed once we remove bootstrap for all type of inputs), so would be better to not duplicate them. But yeah, the model of reusable components you described above sounds good.

There are also differences between radio inputs used in stream settings and org settings since the stream settings have multil-ine label, so would require some CSS changes to have a common class (and thus CSS) for both types of settings or separate class. And also the margin values are different in different cases depending upon the label. Will probably need some tweaking in rules too if we want to avoid having multiple components. Will look into it.

timabbott replied:
I think two separate classes is fine for the two styles -- there's no need to try to fit the same class on two different components. Is the `cursor` property no the default for browsers (you can check looking at the `user-agent` stylesheets)? It's possible it's just an unnecessary line.

sahil839 replied:
The browser value is `cursor: default` and we want `cursor: pointer`.

timabbott commented:
Those focus outlines do look ugly; probably that's replacing their CSS is a question we can post in #design once we've removed the Bootstrap part, and thus changing the CSS is just a localized change.

sahil839 commented:
Updated the PR.
- Updated the commits for stream privacy choices and display settings elements to add a new class and use it to add CSS. For stream settings elements, have added the class to `div.radio-input-parent` element as there is some CSS for the div element as well and for display settings elements added the class to label element since it contains the input element.
- Removed the commit for moving common CSS rules to `zulip.css` for now.
- Added one commit to remove a redundant rule. Though, it could have been skipped in the commit which adds rules to `subscriptions.css` itself but that will break the UI in that commit. So, added it as a new commit after removing rules from `bootstrap.css`.

A couple of points -
- One another approach can be to add the new class to a parent div which contains all the radio elements since we would mostly use radio buttons in a group.
- The `div.radio-input-parent` element can actually be removed by updating some CSS rules, but that can be done as a follow-up.

sahil839 commented on static/styles/subscriptions.css:
```diff
@@ -961,27 +961,43 @@ h4.user_group_setting_subsection_title {
-                margin: 5px;
-                padding-left: 0;
-            }
+    input {
```
Note that `margin` for this `input` is defined separately, since there are different margins for stream creation and edit form. Though this will no longer be required when we remove the modal in stream edit panel as per #23013.

timabbott replied:
Cool, let's try to remember to do that when we merge that PR.

sahil839 commented:
I have started working on removing the bootstrap rules for `textarea` elements as well. So, almost all the textarea elements have different height and/or width, and most of the other properties are same except `margin-bottom` and `border` at couple of places.

Essentially all `textarea` elements in settings sections are also not same due to different height and widths and other than settings the textarea elements are used in compose, message edit, invite and a couple of portico templates. So, even if we create a `settings-textarea` class to use for `textarea` elements in settings sections, then also we would have to duplicate around 13-15 lines in 3-4 files, and those `textarea` elements in settings also will have different widths and heights apart from common rules for `settings-textarea` class.

Do we want to consider adding some common rules for `textarea` elements in `components.css` by making sure none of those are overridden in the rules for individual elements or we are fine with duplication?

timabbott commented:
> Do we want to consider adding some common rules for textarea elements in components.css by making sure none of those are overridden in the rules for individual elements or we are fine with duplication?

Can you post in `#frontend>textarea css reset` about this? I'd like to get Anders' thoughts as well. My general intuition is we should feel free to define a couple CSS classes that we apply one of to all the existing textarea elements and we put in `components.css` or probably better `app_components.css` for any common CSS we want everywhere, but there's also nothing wrong with some duplication if we have a pair of textareas that are not actually intended to be identical components.

timabbott commented on static/templates/stream_settings/stream_types.hbs:
```diff
@@ -4,7 +4,7 @@
     </label>
     {{#each stream_privacy_policy_values}}
-        <div class="radio-input-parent">
+        <div class="radio-input-parent settings-radio-input">
```
Similarly, this is not an `input` element; the previous class name `radio-input-parent` seems better on this front; maybe overall we want `settings-radio-input-parent` (and for there to be only one class on this element, so we can just rename `radio-input-parent` first).

timabbott commented:
I posted a few comments on naming; otherwise this looks good to me (not having tested manually; I assume you've done the before/after `git checkout` verification that it looks the same with or without this branch).

sahil839 commented:
Updated the PR -
- Changed `display-settings-radio-input` class to `display-settings-radio-choice-label`. Have not named it `display-settings-radio-input-parent` since we use `settings-radio-input-parent` is being used for `div` and `choice-label` seemed better. We can though probably remove the containing `div` in stream settings with some modifications in CSS as a follow-up.
- Added commit to rename `radio-input-parent` to `settings-radio-input-parent` and then added bootstrap CSS rules to this new class.

And yes I have tested manually by comparing the UI before and after these changes.

timabbott commented:
Sounds good, merged, thanks @sahil839!

Hopefully the follow-ups for other input elements can go faster as we work out the strategy for these things :)
